typed-racket-lib/typed-racket/typecheck/tc-expr-unit.rkt

```diff
@@ -352,6 +352,25 @@
           #:stx form
           "expected single value, got multiple (or zero) values")]))
 
+;; Apply `single-value` to a list of forms in order, accumulate the prop info
+(define (map-single-value forms [props '()])
+  (define any-res (-tc-any-results #f))
```

Comment:
Would `check-in-order/single-value` (or something similar) be a clearer name? `map-single-value` seems too generic to me.

Reply:
Renamed to `map-single-value/unconditional-prop` because:

- `map single-value` is what we were doing before
- `map` implies left-to-right order
- `/unconditional-prop` explains the 2nd argument
- and the prop that gets accumulated is the "unconditional prop" from the arguments

I'm happy with this name, but not opposed to changing it

Reply:
I've thought more about this and looked at the code more; here are a few comments:

1. If this is a good change and the right thing to do (and I think it is) we should be doing it any time we can, every time we check function arguments. In particular we should also be accumulating the propositions [here](https://github.com/racket/typed-racket/blob/18383003b43adf50aa5af6cb5c1fe92d286eae7e/typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-main.rkt#L98), [here](https://github.com/racket/typed-racket/blob/18383003b43adf50aa5af6cb5c1fe92d286eae7e/typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-main.rkt#L97)., [here](https://github.com/racket/typed-racket/blob/18383003b43adf50aa5af6cb5c1fe92d286eae7e/typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-main.rkt#L91), etc, i.e. we should do it for both standard and dependent functions _whenever_ there could be propositions that help the remaining arguments type check.

2. Does it make sense to also do this on arguments were we have calculated an expected type (i.e. [here](https://github.com/racket/typed-racket/blob/18383003b43adf50aa5af6cb5c1fe92d286eae7e/typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-main.rkt#L111))? I can't remember... is there a way we can have an expected type but still extract meaningful propositions from an expression? If so we should remember those props as well.

3. With these changes, we're now extracting the unconditional propositions from argument type results twice (see [here](https://github.com/racket/typed-racket/blob/18383003b43adf50aa5af6cb5c1fe92d286eae7e/typed-racket-lib/typed-racket/typecheck/tc-funapp.rkt#L64) for the other time). I think we should probably just do this once (i.e. _while_ we're checking the argument types in order from left-to-right as you have done here) and then we should pass that prop along (maybe as an additional argument to [tc/funapp](https://github.com/racket/typed-racket/blob/18383003b43adf50aa5af6cb5c1fe92d286eae7e/typed-racket-lib/typed-racket/typecheck/tc-funapp.rkt#L62)) so it can be propagated/used where needed. In other words, the logic from the code I linked above that is also extracting unconditional argument propositions should appear in whatever we call `map-single-value/unconditional-prop`.

4. I think `map` as a part of the name is a misnomer because it is accumulating something other than a list along the way (the unconditional props) that it should be returning in addition to the list of type results (i.e. we should pass along that accumulated logical info since it is used/needed elsewhere).

5. One other misc note: this behavior reminds me of `tc-body/check` which also accumulates unconditional info as It checks expressions guaranteed to be evaluated in order. This observation may not be useful... but I figured I'd mention it.

Reply:
`2.` yes
`4.` "map" alone would be bad to me, but I like the current "map with prop". (I'm happy to take suggestions, but I'm personally not going to think about new names)